### ISSUE TITLE
Use image_type_to_extension() to get the file extension from the image type

### ISSUE THREAD

**Issue by avpaderno** [backdrop/backdrop-issues#5859]
[`image_gd_get_info()`](https://docs.backdropcms.org/api/backdrop/core%21modules%21system%21image.gd.inc/function/image_gd_get_info/1) hard-codes an array to get the image extension from the image type.

```php
$extensions = array('1' => 'gif', '2' => 'jpg', '3' => 'png');
if (defined('IMAGETYPE_WEBP')) {
  $gd_info = gd_info();
  if (isset($gd_info['WebP Support']) && $gd_info['WebP Support'] == TRUE) {
    // Extend mapping of constants to extensions, as the PHP and GD
    // versions support WebP.
    // @see https://www.php.net/manual/en/image.constants.php
    $extensions['18'] = 'webp';
  }
}
$extension = isset($extensions[$data[2]]) ? $extensions[$data[2]] : '';
```

That isn't anymore necessary: Starting from PHP 5.2.0, [`image_type_to_extension()`](https://www.php.net/manual/en/function.image-type-to-extension.php) gives that information. That code can be simplified, when PHP 5.6 is the minimum required PHP version.

```php
$extension = image_type_to_extension($data[2], FALSE) ?: '';
```

The full function code would become the following.

```php
function image_gd_get_info(stdClass $image) {
  $details = FALSE;
  $data = @getimagesize($image->source);

  if (isset($data) && is_array($data)) {
    $details = array(
      'width' => $data[0],
      'height' => $data[1],
      'extension' => image_type_to_extension($data[2], FALSE) ?: '',
      'mime_type' => $data['mime'],
    );
  }

  return $details;
}
```

**Comment by kiamlaluno** [comment]
Some image tests are falling, but they don't seem tests related to this change.

**Comment by kiamlaluno** [comment]
I found the reason: BackDrop expects the extension for JPEG images to be jpg, but `image_type_to_extension()` always return `'jpeg'`. The tests are checking the image extension and execute the wrong `if`-branch.

I  got confused from the test failure message.

**Comment by indigoxela** [comment]
@kiamlaluno great, many thanks for finding and fixing the flaw. :+1: 

I do have one concern, though... Not sure, how this was meant to be, but the hardcoded extensions array also (to my understanding) limits to extensions appropriate for web usage (jpeg, png, gif, webp).

I may be wrong, though, and that limit shouldn't happen inside image_gd_get_info().

**Comment by kiamlaluno** [comment]
`image_gd_get_info()` should just return information about the images supported by the GD library. I guess the array was just used because that was the only way to get the image extension from the image type returned by `getimagesize()`.

At least, this is what I get from [`image_load()`](https://docs.backdropcms.org/api/backdrop/core%21includes%21image.inc/function/image_load/1), which first calls `image_gd_get_info()` (or the equivalent function for the image toolkit set by the user) and then call[s `image_gd_load()`](https://docs.backdropcms.org/api/backdrop/core%21modules%21system%21image.gd.inc/function/image_gd_load/1).

**Comment by kiamlaluno** [comment]
There are two test methods that fails. I checked their code, but I don't see anything related to images, this time.  
I  closed and re-opened my PR to make the tests run again. Let's see if they fail again.

**Comment by kiamlaluno** [comment]
All the checks are now passing.

**Comment by indigoxela** [comment]
This looks great now, but ...

For GD lots of types are "images", for example xbm. Does for example [file_validate_is_image()](https://docs.backdropcms.org/api/backdrop/core%21includes%21file.inc/function/file_validate_is_image/1) lead to unexpected results then?

To be clear: I'm totally happy that another place where hardcoded values get cleaned up, but the cleanup might lead to changed behavior in other functions. In core it doesn't lead to any problems, though, so maybe I'm just paranoid. :wink:

So... I'm tentatively leaning towards RTBC, but would be glad if another pair of eyes could verify.

**Comment by kiamlaluno** [comment]
I agree: We need to verify Backdrop core doesn't wrongly assume the image type recognized by the image toolkit.

If `file_validate_is_image()` is supposed to effectively validate only three image types, its code should be changed, as it accepts every image types for which the image toolkit returns an extension. In the case of the GD image toolkit, those image types are actually four.

**Comment by kiamlaluno** [comment]
I tried playing with the Tugboat preview created for the PR, but GD library used by the server doesn't support the webp format. I will try with a different image format.

**Comment by indigoxela** [comment]
Other file upload candidates, that don't use extra extension validation ... :thinking:

- The traditional "User picture", which gets activated on /admin/config/people/settings
- Possibly the logo upload ... not sure: /admin/config/system/site-information
- Eventually a custom file type? /admin/structure/file-types/add (not sure if that's relevant)

**Comment by kiamlaluno** [comment]
User pictures and logo cannot be a .webp image. I haven't found a place where Backdrop accepts every image handled by the library used from the enabled image toolkit.
So far, it seems the PR change isn't going to cause issues to Backdrop and lead it to accept extra image types.

**Comment by kiamlaluno** [comment]
Given [the functions/methods that calls `image_load()`](https://docs.backdropcms.org/api/backdrop/core%21includes%21image.inc/function/calls/image_load/1), I guess that the chance that Backdrop core could accept all the images handled by the library used from the image toolkit are very low.

**Comment by avpaderno** [comment]
I am closing this issue as outdated because other changes that have been done in other issues.
